hw/misc/pcie-testdev.c

```diff
@@ -0,0 +1,202 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "qemu/osdep.h"
+#include "hw/pci/pci.h"
+#include "hw/qdev-properties.h"
+#include "sysemu/kvm.h"
+
+#define PCI_EXPRESS_DEVICE_ID_TEST 0x0010
+#define TYPE_PCIE_TEST_DEV "pcie-testdev"
+
+#define SRCADDR 0x00
+#define DSTADDR 0x08
+#define SIZE 0x10
+#define DOORBELL 0x18
+#define STATUS 0x20
+#define BUSY_DELAY 0x28 /* delay in ms to simulate a longer transfer */
+
+typedef struct DMARegs {
+    dma_addr_t srcaddr;
+    dma_addr_t dstaddr;
+    dma_addr_t size;
+    dma_addr_t doorbell;
+    dma_addr_t status;
+    dma_addr_t busy_delay;
+} DMARegs;
+
+typedef struct PCIeTestDevState {
+    PCIDevice parent_obj;
+    MemoryRegion membar;
+    DMARegs regs;
+    bool aggressive;
+    bool aggression_delay; /* delay between aggressive DMA transfers */
+} PCIeTestDevState;
+
+#define PCI_EXPRESS_TEST_DEV(obj) \
+    OBJECT_CHECK(PCIeTestDevState, (obj), TYPE_PCIE_TEST_DEV)
+
+static Property props[] = {
+    DEFINE_PROP_BOOL("aggressive", PCIeTestDevState, aggressive, false),
+    DEFINE_PROP_UINT64("aggression_delay_ms", PCIeTestDevState, aggression_delay, false),
+    DEFINE_PROP_UINT64("srcaddr", PCIeTestDevState, regs.srcaddr, false),
+    DEFINE_PROP_UINT64("dstaddr", PCIeTestDevState, regs.dstaddr, false),
+    DEFINE_PROP_UINT64("size", PCIeTestDevState, regs.size, false),
+    DEFINE_PROP_UINT64("busy_delay_ms", PCIeTestDevState, regs.busy_delay, false),
+    DEFINE_PROP_END_OF_LIST()
+};
+
+static uint64_t
+pcie_testdev_read(void *opaque, hwaddr addr, unsigned size)
+{
+    PCIeTestDevState *s = opaque;
+
+    switch(addr) {
+    case SRCADDR:
+        return s->regs.srcaddr;
+        break;
+    case DSTADDR:
+        return s->regs.dstaddr;
+        break;
+    case SIZE:
+        return s->regs.size;
+        break;
+    case STATUS:
+        return s->regs.status;
+        break;
+    case BUSY_DELAY:
+        return s->regs.busy_delay;
+    default:
+        return 0x0;
+    }
+}
+
+static void
+enable_bus_master(PCIeTestDevState *s){
+    pci_set_word(s->parent_obj.config + PCI_COMMAND,
+                 s->parent_obj.config[PCI_COMMAND] | PCI_COMMAND_MASTER);
+}
+
+static void
+dma_transfer(PCIeTestDevState *s){
+    s->regs.status = 0x01;
+    printf("DMA Transfer from source addr %p to dest addr %p, length %lu\n", (void *)s->regs.srcaddr, (void *)s->regs.dstaddr, s->regs.size);
+    uint16_t *buf = g_malloc(s->regs.size);
```

Comment:
Might want to use g_malloc0 to prevent unitialized memory from leaking into the destination buffer.

Also, please check for errors and don't write to the destination if the read from the source failed.

Reply:
Used g_malloc0, only write when read succeeds. Trace event records overall transfer status.